ribs/emitters/opt/_openai_es.py

```diff
@@ -0,0 +1,191 @@
+
+    Refer to :class:`EvolutionStrategyBase` for usage instruction.
+
+    Args:
+        sigma0 (float): Initial step size.
+        batch_size (int): Number of solutions to evaluate at a time. If None, we
+            calculate a default batch size based on solution_dim.
+        solution_dim (int): Size of the solution space.
+        seed (int): Seed for the random number generator.
+        dtype (str or data-type): Data type of solutions.
+        mirror_sampling (bool): Whether to use mirror sampling when gathering
+            solutions. Defaults to true.
+        weight_rule (str): Not used.
+        adam_kwargs (dict): Keyword arguments passed to :class:`AdamOpt`.
+    """
+
+    def __init__(
+            self,
+            sigma0,
+            solution_dim,
+            batch_size=None,
+            seed=None,
+            dtype=np.float64,
+            weight_rule="truncation",
+            mirror_sampling=True,
+            **adam_kwargs):
+        super().__init__(
+            sigma0,
+            solution_dim,
+            batch_size,
+            seed,
+            dtype,
+        )
+        self.weight_rule = weight_rule  # Not used.
+        self.mirror_sampling = mirror_sampling
+
+        # Default batch size should be an even number for mirror sampling.
+        if batch_size is None and self.batch_size % 2 != 0:
+            self.batch_size += 1
+
+        if self.batch_size <= 1:
+            raise ValueError("Batch size of 1 is not supported because rank"
+                             " normalization does not work with batch size of"
+                             " 1.")
+
+        if self.mirror_sampling and self.batch_size % 2 != 0:
+            raise ValueError("If using mirror sampling, batch_size must be an"
+                             " even number.")
+
+        # Strategy-specific params -> initialized in reset().
+        self.adam_opt = AdamOpt(self.solution_dim, **adam_kwargs)
+        self.last_update_ratio = None
+        self.noise = None
+
+    def reset(self, x0):
+        """Resets the optimizer to start at x0.
+
+        Args:
+            x0 (np.ndarray): Initial mean.
+        """
+        self.adam_opt.reset(x0)
+        self.last_update_ratio = np.inf  # Updated at end of tell().
+        self.noise = None  # Becomes (batch_size, solution_dim) array in ask().
+
+    def check_stop(self, ranking_values):
+        """Checks if the optimization should stop and be reset.
+
+        Args:
+            ranking_values (np.ndarray): Array of objective values of the
+                solutions, sorted in the same order that the solutions were
+                sorted when passed to ``tell()``.
+
+        Returns:
+            True if any of the stopping conditions are satisfied.
+        """
+        if self.last_update_ratio < 1e-9:
+            return True
+
+        # Fitness is too flat (only applies if there are at least 2
+        # parents).
+        if (len(ranking_values) >= 2 and
```

Comment:
See `check_stop` in CMA-ES -- we use np.linalg.norm there because there may be multiple ranking values.